java-client/session/src/main/java/io/deephaven/client/impl/SessionImpl.java

```diff
@@ -0,0 +1,342 @@
+        executor.schedule(SessionImpl.this::refreshSessionToken, refreshDelayMs,
+            TimeUnit.MILLISECONDS);
+    }
+
+    private void scheduleRefreshSessionTokenNow() {
+        executor.schedule(SessionImpl.this::refreshSessionToken, 0, TimeUnit.MILLISECONDS);
+    }
+
+    private void refreshSessionToken() {
+        HandshakeRequest handshakeRequest = HandshakeRequest.newBuilder().setAuthProtocol(0)
+            .setPayload(ByteString.copyFromUtf8(auth.session())).build();
+        HandshakeHandler handler = new HandshakeHandler();
+        sessionService.refreshSessionToken(handshakeRequest, handler);
+    }
+
+
+    private static class PublishObserver
+        implements ClientResponseObserver<BindTableToVariableRequest, BindTableToVariableResponse> {
+        private final CompletableFuture<Void> future = new CompletableFuture<>();
+
+        @Override
+        public void beforeStart(
+            ClientCallStreamObserver<BindTableToVariableRequest> requestStream) {
+            future.whenComplete((session, throwable) -> {
+                if (future.isCancelled()) {
+                    requestStream.cancel("User cancelled", null);
+                }
+            });
+        }
+
+        @Override
+        public void onNext(BindTableToVariableResponse value) {
+            future.complete(null);
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            future.completeExceptionally(t);
+        }
+
+        @Override
+        public void onCompleted() {
+            if (!future.isDone()) {
+                future.completeExceptionally(
+                    new IllegalStateException("Observer completed without response"));
+            }
+        }
+    }
+
+
+    private class SessionCallCredentials extends CallCredentials {
+
+        @Override
+        public void applyRequestMetadata(RequestInfo requestInfo, Executor appExecutor,
+            MetadataApplier applier) {
+            AuthenticationInfo localAuth = auth;
+            Metadata metadata = new Metadata();
+            metadata.put(Key.of(localAuth.sessionHeaderKey(), Metadata.ASCII_STRING_MARSHALLER),
+                localAuth.session());
+            applier.apply(metadata);
+        }
+
+        @Override
+        public void thisUsesUnstableApi() {
+
+        }
+    }
+
+    private class HandshakeHandler implements StreamObserver<HandshakeResponse> {
+
+        @Override
+        public void onNext(HandshakeResponse value) {
+            auth = AuthenticationInfo.of(value);
+            scheduleRefreshSessionToken(value);
+            handler.onRefreshSuccess();
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            handler.onRefreshTokenError(t, SessionImpl.this::scheduleRefreshSessionTokenNow);
+        }
+
+        @Override
+        public void onCompleted() {
+            // ignore
+        }
+    }
```

Comment:
The obligatory this is going to better match arrow flight. Just FYI; change is incoming on how this works.

Reply:
I'm not sure what you mean here.